### PR TITLE
[platform_tests] add test to make sure cpuidle is off for m0 and mx

### PR DESCRIPTION
**Pull request by Xichen96**
### Description of PR

Currently some Intel CPU tends to have problem when CPU enters idle state. We expect Intel CPUs to not have an active idle driver and idle states.

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012
- [x] 202205

#### What is the motivation for this PR?
Make sure Intel CPU m0 won't enter idle states.

#### How did you do it?
Check active idle driver (should be none) and available idle states (should be none)

#### How did you verify/test it?
Run test on device

#### Any platform specific information?
Intel m0 only

### CODE REVIEW

File: tests/platform_tests/test_idle_driver.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+"""
+import logging
+import pytest
+
+from tests.common.helpers.assertions import pytest_assert
+
+logger = logging.getLogger(__name__)
+
+pytestmark = [
+    pytest.mark.topology('m0'),
```
Comment (lizhijianrd) [inline tests/platform_tests/test_idle_driver.py:14]: Should we enable this testcase on `mx` and other topology?

File: tests/platform_tests/test_idle_driver.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+]
+
+def test_idle_driver(duthosts, enum_rand_one_per_hwsku_hostname):
+    duthost = duthosts[enum_rand_one_per_hwsku_hostname]
+    is_intel = 'Intel' in duthost.shell('lscpu | grep Vendor')['stdout']
```
Comment (Blueve) [inline tests/platform_tests/test_idle_driver.py:19]: AMD CPU also have c state defined. 
By checking this thread, we can use same approach to check:
https://askubuntu.com/questions/1242503/how-do-you-check-if-processor-max-cstate-is-set-correctly-using-terminal-tilix

File: tests/platform_tests/test_idle_driver.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    if not is_intel:
+        pytest.skip('CPU vendor is not Intel')
+    idle_driver = duthost.shell('cat /sys/devices/system/cpu/cpuidle/current_driver')['stdout']
+    pytest_assert(idle_driver == 'none', 'Idle driver {} is active on this m0 device'.format(idle_driver))
```
Comment (Blueve) [inline tests/platform_tests/test_idle_driver.py:23]: I think either idle driver disabled or max c-state = 1 are acceptable

File: tests/platform_tests/test_idle_driver.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    cpuidle = duthost.command('ls /sys/devices/system/cpu/cpu0/cpuidle/', module_ignore_errors=True)['rc']
+    pytest_assert(cpuidle == 2, "cpuidle directory should not exist: rc {}".format(cpuidle))
```
Comment (Blueve) [inline tests/platform_tests/test_idle_driver.py:25]: If max C state is C1, then it is acceptable as well

File: tests/platform_tests/test_idle_driver.py (hunk 1 above)
Comment (Blueve) [reply tests/platform_tests/test_idle_driver.py:14]: Echo on this. We need enable for Mx as well. Please update the header as well

File: tests/platform_tests/test_idle_driver.py (hunk 3 above)
Comment (Xichen96) [reply tests/platform_tests/test_idle_driver.py:23]: do you mean max_cstate=1 for intel_idle?

File: tests/platform_tests/test_idle_driver.py (hunk 2 above)
Comment (Xichen96) [reply tests/platform_tests/test_idle_driver.py:19]: If we do not have the same device hang problem on amd cpu, do we need to worry about it?

File: tests/platform_tests/test_idle_driver.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+]
+
+def test_idle_driver(duthosts, enum_rand_one_per_hwsku_hostname):
+    duthost = duthosts[enum_rand_one_per_hwsku_hostname]
+    is_intel = 'Intel' in duthost.shell('lscpu | grep Vendor')['stdout']
+    if not is_intel:
+        pytest.skip('CPU vendor is not Intel')
+    idle_driver = duthost.shell('cat /sys/devices/system/cpu/cpuidle/current_driver')['stdout']
+    pytest_assert(idle_driver == 'none', 'Idle driver {} is active on this m0 device'.format(idle_driver))
+    cpuidle = duthost.command('ls /sys/devices/system/cpu/cpu0/cpuidle/', module_ignore_errors=True)['rc']
```
Comment (Blueve) [inline tests/platform_tests/test_idle_driver.py:24]: check all cpus

File: tests/platform_tests/test_idle_driver.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+"""
+Some devices have potential problems entering idle state. So
+on m0 devices, we expect to disable both intel idle driver
```
Comment (Blueve) [inline tests/platform_tests/test_idle_driver.py:3]: Please update this statement, just remove "So on m0 devices".

**Review by Blueve — APPROVED** [review]

**Review by Blueve — APPROVED** [review]

**Comment by AharonMalkin** [comment]
Hey @Xichen96  @yxieca @Blueve 
Is this test relevant only for m0 topo? If yes, please explain why.

Thank you

**Comment by Xichen96** [comment]
@AharonMalkin Some Intel CPUs are found to have caused problems when they enter certain idle state. We disable idle driver as a precaution. We start from M0 and Mx because they are weaker devices with less implecations. Ideally, we wouldn't want our management devices to sleep at all.

**Comment by AharonMalkin** [comment]
@Xichen96 Why not enable this test for all topologies?  What prevents it to happen on t0 for example?

**Pull request merged**